### DISCUSSION TITLE
Light-weight use of Quarto extensions

### DISCUSSION THREAD

**Discussion by allefeld** [quarto-dev/quarto-cli#9091] [Feature Requests]
### Description

I proposed this idea [elsewhere](https://github.com/quarto-dev/quarto-cli/discussions/8850#discussioncomment-8731083), but it didn't receive a response, so I'm posting it as a separate feature request. It is related to the comments https://github.com/quarto-dev/quarto-cli/discussions/2845#discussioncomment-7872412 and https://github.com/quarto-dev/quarto-cli/discussions/2845#discussioncomment-8613469, but as far as I can see there is no equivalent existing feature request.

### Background & Motivation

Quarto has [no global configuration](#8850), which means that if a user has a certain preferred way to set things up (options, stylesheets, etc.), they have to copy over this configuration into each new project. Or they can create an extension for a custom format, which then has to be installed in each new project – but this basically does the same copying, albeit in a neater way. Consistent with no global configuration in general, extensions cannot be installed "into Quarto", only into a specific project. That is a [design decision which will not be changed](https://github.com/quarto-dev/quarto-cli/discussions/8850#discussioncomment-8561620), based on the goal to have reproducibly rendered, self-contained projects.

The problem with that is that

> there are many use-cases for Quarto which do not fit into the self-contained reproducible project mold.
>
> For example, I use Quarto to write solutions for exercises in a module I teach. In my module, there are different contents for each week, and when the module repeats the next year there may be changes and therefore new versions. Each solution fits into a single qmd file – it's not a project and doesn't intend to be. Now I can put all my settings into the YAML metadata block, but that means I have to copy-and-paste for every qmd file I create. And maybe at some point I decide to change something, and then I have to go back to every file and change it there. I had hoped that custom formats would solve this, but then files get copied into each weekly / yearly folder, cluttering them up. And when I change the format, I have to run quarto update everywhere.

@go-bayes commented

>   @cscheid, respectfully, many users want to update LaTeX macros, along with bibliographies, functions, templates, and much else, across their projects. Duplication is not only inefficient, it leads to errors. And indeed, when we use PDF we can obtain this desired behaviour using header-includes: - \input{/home/.../macro-file-here.tex}. This is, at any rate, my use case. With @allefeld
>   
>   >   I'd rather not have dozens of copies of my math-definitions.tex.

Defining an extension for a custom format makes things slightly better, because at least now the clutter goes into a subdirectory `_extensions`, but it doesn't change the fact that lots of files are duplicated (triplicated, …) and that a simple single-file document carries with it a whole bunch of ancillary files. It would be perfect if an extension could be installed "into Quarto" instead of a project, but that goes against the "no global configuration" design decision which will not be changed.

### Proposal "Light-weight use of Quarto extensions"

>   Make "installing" an extension something that can be achieved by a YAML metadata key. E.g.
>   
>   ```yaml
>   ---
>   title: "Presentation"
>   subtitle: "LexConf 2022"
>   author: "Your Name"
>   date: today
>   use-extension: lexcorp/lexconf
>   format: lexconf-revealjs
>   ---
>   ```
>   
>   The extension / custom format would be hosted on GitHub etc., if needed with version management. The contents are not downloaded into the directory of the document, but cached by Quarto.

This would satisfy both Quarto's focus on reproducibility and my and some other users' wish for reduced clutter and duplication (because the extension files are not put into the folder of the file). For full reproducibility, the user would specify a particular version, e.g. `lexcorp/lexconf@v1.2`, or they could just use the generic reference `lexcorp/lexconf` and always get the latest version. Both is already possible with installing an extension into a project, but clutter & duplication are avoided.

**Comment by allefeld** [comment]
Checking in: Is there any chance to get such a feature?
In case I tried to implement it myself: Is there a chance that it gets adopted (merged)?

**Reply by mcanouil** [reply]
There is something I don't get with your proposal: where are stored the extensions?

> The contents are not downloaded into the directory of the document, but cached by Quarto.

How this changed anything? Quarto "cache" is `.quarto` is project specific.
Extensions still needs to be downloaded, and are still in your project.
Also, on CI, this means, you download every single time the extensions.


You said, your use case does not fit into a project, but did you consider using the `default project type to share extensions across your "teaching modules"?


Edit: see <https://github.com/mcanouil/quarto-issues-experiments/tree/main/test-quarto-shared-extensions>.

````sh
.
├── README.md
├── _extensions
│   └── mcanouil
│       ├── iconify
│       │   ├── LICENSE
│       │   ├── _extension.yml
│       │   ├── iconify-icon.min.js
│       │   └── iconify.lua
│       └── preview-colour
│           ├── LICENSE
│           ├── _extension.yml
│           └── preview-colour.lua
├── _quarto.yml
├── module1
│   ├── module1.html
│   └── module1.qmd
└── module2
    ├── module2.html
    └── module2.qmd

7 directories, 13 files
````

**Reply by allefeld** [reply]
I didn't mean Quarto's `.cache` directory within the project directory, but "cached" as a general term. That is usually implemented using a user-specific or system-wide directory, on Linux e.g. `$XDG_CACHE_HOME/quarto/` or `/var/cache/quarto/`. On my computer, the former already exists as `$HOME/.cache/quarto/` and contains files related to SASS processing among others.

I wasn't aware of your Shared Extensions Experiment, it's not linked to from the Quarto documentation, but I will have a look.

**Reply by mcanouil** [reply]
> I wasn't aware of your Shared Extensions Experiment, it's not linked to from the Quarto documentation, but I will have a look.

I made it to showcase this to you.
There is nothing very advanced in it once you understand what a Quarto project is.

**Reply by allefeld** [reply]
Thanks for creating the example, I now looked it over.

What I understand is that instead of installing extensions into each directory containing a document, they are installed into a higher up project directory. That makes sense and reduces clutter if directories are organised this way (which for me is often the case, though not always).

What I don't understand is what `project: type: default` does. I have used `website` projects, and it is relatively obvious what a `book` or `manuscript` project is – but what is a `default` project? I've searched the documentation but didn't come up with anything. [Hugo Example](https://quarto.org/docs/extensions/project-types.html#hugo-example) uses `project: type: default`, but there is no explanation.

Generally, the documentation of project types is a bit sparse, or maybe just not well organised. On the [Project Basics](https://quarto.org/docs/projects/quarto-projects.html) page in the *Guide* there is a list of project types in the form of output from `quarto create project`. *Websites*, *Manuscripts*, and *Books* have dedicated entries  menu on the left, but the others don't. And in the *Reference* part of the documentation, *Projects* has sections again on *Websites*, *Books*, and *Manuscripts*, but not the other types.

**Reply by mcanouil** [reply]
Blog are described in Website because it's simply a website plus a listing.
Whenever you don't have the other types, it's default. When you don't set anything, it's default. Briefly default is default.

**Reply by allefeld** [reply]
Okay, but that still doesn't explain what it does.

**Reply by mcanouil** [reply]
I don't understand what you want/expect here.
Everything in Quarto is a project (especially since 1.5). A single document is a project where the document defines the root and is a "default" project. Setting a `_quarto.yml` means you want your directory to be a project and the root of that project is where the `_quarto.yml` is, by default the project is "default", even if you did not set it explicitly.

**Reply by allefeld** [reply]
What I want/expect is that the documentation explains how the software works.

Nowhere in the documentation is it stated that "everything in Quarto is a project". The *Guide* says:

> Quarto projects are directories that provide:
> 
> - A way to render all or some of the files in a directory with a single command (e.g. `quarto render myproject`).
> 
> - A way to share YAML configuration across multiple documents.
> 
> - The ability to redirect output artifacts to another directory.
> 
> - The ability to freeze rendered output (i.e. don’t re-execute documents unless they have changed).
> 
> - In addition, projects can have special “types” that introduce additional behavior (e.g. [websites](https://quarto.org/docs/websites/) or [books](https://quarto.org/docs/books/)).

That strongly suggests that Quarto works with two kinds of input, a) single files where these additional features are not available (or do not apply) and b) directories containing multiple files (aka projects) where these additional features are available.

Moreover, all the project types which are documented (*Websites*, *Books*, and *Manuscripts*) create some sort of integration, where output files are either connected through links (e.g. in the navbar or sidebar) or there is only a single output file. That strongly suggests that such integration is the point of projects.

The use case I explained above does not call for such an integration; I want each single file treated as a single file, just sharing some configuration, which I assumed should be done via custom formats, which then motivated my proposal for lightweight use of such custom formats.

After all this back and forth, I suspect a `default` project is one where there is no such integration – is that correct?

If yes, then it does indeed partially solve my problem – but I still think that my lightweight use proposal makes sense.

**Reply by mcanouil** [reply]
> But if a user tells you that from the documentation they don't understand something, then you should consider the possibility that the documentation needs to be improved.

I don't believe, me not understanding what you are missing (clearly stated/written that way, that is even the first sentence of https://github.com/quarto-dev/quarto-cli/discussions/9091#discussioncomment-9175207) is equivalent to me dismissing the possibility to improve the documentation.
Please refrain to make assumption of me not considering the possibility to improve the documentation.
That's a personal attack which is against the [Code of Conduct](https://github.com/quarto-dev/quarto-cli?tab=coc-ov-file).

I/We created an issue template precisely to improve the documentation based on user's feedback: https://github.com/quarto-dev/quarto-cli/issues/new?assignees=&labels=documentation&projects=&template=documentation.yml

**Reply by allefeld** [reply]
I deleted my offensive comment.

**Comment by baptiste** [comment]
[Just curious](https://github.com/quarto-dev/quarto-cli/discussions/2845#discussioncomment-8738269), can you not [make softlinks from a centralised location?](https://github.com/baptiste/test-global)

**Reply by mcanouil** [reply]
@baptiste When using text with hyperlinks, the text needs to explain what the link is about.
Currently, your comment is not at all accessible (even without disabilities, it's not meaningful as is), see <https://www.w3.org/WAI/tips/writing/#make-link-text-meaningful>.
